### PR TITLE
Adding ExtractAllLocationUri's visitor

### PR DESCRIPTION
**Pull request by nickmarston**
Adds a visitor to create a HashSet of every ArtifactLocation that is present in a SARIF file and resolves those URIs to their index location.

This will be needed for bug filing, where the URIs in a SARIF location will be used as the key to indexing services.  I will next be looking at the best place to transform this artifact location list into a list of unique URIs, normalized, which I can then pass to the needed service.

### CODE REVIEW

File: src/Test.Utilities.Sarif/TestConstants.cs [hunk 1]
```diff
@@ -28,6 +28,14 @@ public static class RuleIds
             public const string Rule10 = "TST0010";
         }
 
+        public static class FileLocations
+        {
+            public const string Location1 = "C:\\Test\\Data\\File1.sarif";
```
Comment (lgolding) [inline src/Test.Utilities.Sarif/TestConstants.cs]: >" [](start = 44, length = 1)

Nit: consider using verbatim constants `@"C:\Test\Data\File.sarif"`.

File: src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+{
+
+    public class ExtractAllArtifactLocationsVisitor : SarifRewritingVisitor
+    {
+        private Run _currentRun;
+        public HashSet<ArtifactLocation> allArtifactLocations { get; private set; }
+
+        public ExtractAllArtifactLocationsVisitor()
+        {
+            allArtifactLocations = new HashSet<ArtifactLocation>(ArtifactLocation.ValueComparer);
```
Comment (lgolding) [inline src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs]: >allArtifactLocations [](start = 12, length = 20)

It's not clear to me whether you want to reset this at the start of each run (at the start of `VisitRun`) or accumulate across runs, as the code stands. You'll need to consider the expected use case, and whatever the answer, you'll need to document it in the class summary comment.

File: src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+
+    public class ExtractAllArtifactLocationsVisitor : SarifRewritingVisitor
+    {
+        private Run _currentRun;
+        public HashSet<ArtifactLocation> allArtifactLocations { get; private set; }
```
Comment (lgolding) [inline src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs]: >a [](start = 41, length = 1)

Convention: All our `public` property names are CamelCase.

File: src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            };
+
+            var visitor = new ExtractAllArtifactLocationsVisitor();
+            visitor.VisitSarifLog(sarifLog);
+
+            foreach (Location location in sarifLog.Runs[0].Results[0].Locations)
+            {
+                visitor.allArtifactLocations.Contains(location.PhysicalLocation.ArtifactLocation).Should().BeTrue();
+            }
+        }
```
Comment (lgolding) [inline src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs]: >} [](start = 8, length = 1)

Do you want to accumulate URIs that occur in other places, such as `Result.RelatedLocations`? That's what the code does today. If you don't want that, the visitor will need to maintain state ("I'm in `Result.Locations`!"). Again, it depends on your use case, and again, either way, you'll need to document the decision. Keep in mind that this is a general purpose SDK; what's right your your use case might not be appropriate in general.

File: src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs (hunk 2 above)
Comment (nickmarston) [reply src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs]: Currently, I expect to get all locations in a given set of runs.  I'll add the documentation confirming this.  I'd be curious if you see the reset version being the more common use case, outside of the bug filer.

---
In reply to: [386741455](https://github.com/microsoft/sarif-sdk/pull/1794#discussion_r386741455) [](ancestors = 386741455)

File: src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs (hunk 4 above)
Comment (nickmarston) [reply src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs]: I tend to think the more useful SDK piece is to extract everything.  Downstream, the filer is going to need to take these extracted locations and filter on them in some way.  Given, however, that ArtifactLocation itself doesn't seem to contain any of the data necessary to make that decision as to where the node came from, it sounds like I may need to store that data.

I need to think on this a bit.  The filer definitely only cares about the locations extracted from the Results.Locations collection, as those are what we want to resolve into area paths for bugs.

I obviously can change the visitor to only hit the nodes I care about (probably at the Results.Location level), but I'm not sure that fits as well.

---
In reply to: [386742798](https://github.com/microsoft/sarif-sdk/pull/1794#discussion_r386742798) [](ancestors = 386742798)

File: src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    /// <summary>
+    /// A visitor that examines a specified SARIF log to extract all artifact locations for later processing.
+    /// This will extract every location present in the log file, which may then need to be filtered by a
+    /// downstream consumer.
+    /// 
+    /// May optionally be based a run at contsruction, which allows the visitor to later visit only specific parts of
```
Comment (michaelcfanning) [inline src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs]: 'be configured by'

File: src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    /// the sarif log where location data aggregation is desired.
+    /// </summary>
+    public class ExtractAllArtifactLocationsVisitor : SarifRewritingVisitor
+    {
+        private Run _currentRun;
```
Comment (michaelcfanning) [inline src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs]: >_currentRun [](start = 20, length = 11)

Suggestion: convert this to a writable public property, CurrentRun. Remove the ctor. Set this property on VisitRun. Now users can set this and visit interior items. But the class always stays consistent. Don't forget to update the class summary if you accept this suggestion.

File: src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+        public ExtractAllArtifactLocationsVisitor(Run node)
+        {
+            AllArtifactLocations = new HashSet<ArtifactLocation>(ArtifactLocation.ValueComparer);
+            _currentRun = node;
+        }
+
+        public override ArtifactLocation VisitArtifactLocation(ArtifactLocation node)
+        {
+            if (node.Uri == null)
```
Comment (michaelcfanning) [inline src/Sarif/Visitors/ExtractAllArtifactLocationsVisitor.cs]: >node [](start = 16, length = 4)

Ditch this conditional. An artifact location may only include an index, which points to the artifact table where the non-null URI is. Add a unit test for this case!

File: src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+        [Fact]
+        public void ExtractAllArtifactLocationsVisitor_ExtractsMultipleLocations()
+        {
+            var sarifLog = new SarifLog
+            {
+                Runs = new[]
+                {
+                    new Run
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs]: >Run [](start = 24, length = 3)

Add an artifacts table with one item at least. Add a result location that refers to this only by index. Add a second case that has both an index and URI.

**Pull request merged**